icicle/appUtils/poseidon/poseidon.cu

```diff
@@ -0,0 +1,256 @@
+#include "poseidon.cuh"
+
+// Used in matrix multiplication
+extern __shared__ scalar_t shared_states[];
+
+__global__ void prepare_poseidon_states(scalar_t * inp, scalar_t * states, size_t number_of_states, scalar_t domain_tag, const PoseidonConfiguration config) {
+    int idx = (blockIdx.x * blockDim.x) + threadIdx.x;
+    int state_number = idx / config.t;
+    if (state_number >= number_of_states) {
+        return;
+    }
+    int element_number = idx % config.t;
+
+    scalar_t prepared_element;
+
+    // Domain separation
+    if (element_number == 0) {
+        prepared_element = domain_tag;
+    } else {
+        prepared_element = inp[state_number * (config.t - 1) + element_number - 1];
+    }
+
+    // Add pre-round constant
+    prepared_element = prepared_element + config.round_constants[element_number];
+
+    // Store element in state
+    states[idx] = prepared_element;
+}
+
+__device__ __forceinline__ scalar_t sbox_alpha_five(scalar_t element) {
+    scalar_t result = element * element;
```

Comment:
Using `scalar_t::sqr` here would be more optimal (although rn `sqr` just redirects to multiplication, in near future it will be improved).

Reply:
Fixed in 24222badb6d0db93bc0f0645ef7103beb4a6fa30